src/sql.rs

```diff
@@ -190,6 +190,30 @@ impl Sql {
         self.with_conn(|conn| conn.query_row(sql.as_ref(), params, f).map_err(Into::into))
     }
 
+    /// Execute a query which is expected to return zero or one row.
+    pub fn query_row_option<T, P, F>(
```

Comment:
I was going to bikeshed the name (I don't like the type in there) and while doing so argued that we should really only have one function.  I think the API of rusqlite is pretty good: https://docs.rs/rusqlite/0.22.0/rusqlite/struct.Statement.html#method.query_row  They have an `OptionalExtension` trait which adds a `Result::optional` method which handles the no row error and turns it into an option.  I haven't actually tried this, but looking at the implementation I think you might be able to import the trait and use it directly even on our wrapper.

Reply:
`query_row` returns `crate::sql::Result` instead of `rusqlite::Result`, so `optional` can't be used on it. And it is generic, so can't be implemented in `sql` module (compiler says "note: only traits defined in the current crate can be implemented for a type parameter"). Also `optional` doesn't handle `NULL`s.

Reply:
I have squashed some commits and renamed the function into `query_row_optional`, but using `OptionalExtension` requires some more refactoring, like getting rid of `with_conn` which may return errors other than `rusqlite::Error`.

Reply:
Fair enough, thanks for looking into it.  I'd have gone with `query_optional_row` but whatever, the bikeshed will have some colour.